adr/0011-python-application-deployment.md

```diff
@@ -0,0 +1,77 @@
+projects run on in production.
+
+Python projects require a Python runtime and Python packages can be installed system-wide or locally using a tool called
+'virtualenv', now officially part of Python3. Dependency and runtime management has become complicated in the move from
+Python2 to Python3 and the introduction of new language features and dependency management tools; the OS distributions 
+used in production use a fixed version of the Python runtime, often with features missing or mutually exclusive with 
+other features (Ubuntu + Python3 + pip + venv for example). Attempts to accommodate the latest runtimes has introduced
+instability and complexity in the rest of the system.
+
+Containerisation has made describing and building Python applications much easier for a project's author and the line 
+between traditional code development and code deployment is less clear. Container descriptions often live within the 
+code repository instead of in a separate one. Containers serve a single purpose, running a single process, but are 
+composable so that a cluster of containers may be described: one for the application, another for a webserver, another
+for a database and so on.
+
+Deploying a container that has the project within perfectly installed and configured is very similar to deploying a fat
+application binary: a single discrete artifact with all dependencies accounted for. Deployment consists of dropping the
+artifact into the runtime and reloading a service.
+
+Because container descriptions live alongside the codebase and are used during development they suffer the same 
+problems as the author-preferred utility scripts/glue code. There are any number of ways to build a container, does 
+installation/testing/linting/etc happen within the container or without? Are the build commands invoked in the same way 
+across projects? If we insist on containers, then is it the project author's responsibility to now maintain a suite of 
+them? Are they the best party to attempt it.
+
+## Decision
+
+Deploy all Python projects within containers.
+
+Testing happens within project containers.
+
+Standardise building of containers.
+
+## Consequences
+
+Testing becomes less hazardous as fewer variables are in play.
+
+Deployment becomes simpler and much of a project's formula can be removed.
+
+Deployment becomes uniform and new projects can benefit from the tooling built for other Python projects.
+
+Adhoc and idiomatic utility scripts are replaced with a single script that encapsulates build and test commands.
+
+Project authors will have a reproducible environment much closer to production and much less unwieldy than a 
+Vagrant virtual machine.
+
+Project releases have a reference environment that non-eLife parties can use (that doesn't require heavy dependencies like Builder and Vagrant).
+
+Project authors (hopefully) feel more agency over how their code is used in production.
```

Comment:
Is the word `agency` a typo here?

Reply:
I think this would be incorrect then as the purpose is to standardise projects.

Reply:
I see your point, if you are standardizing everything then there's not much agency e.g. choosing a new tool

Reply:
Agree, although I can see the sentiment - it gives a feeling of control or level of engagement in the way the code is used in production.

Reply:
A developer is going to know what is best for their code and can express this through the popular Dockerfile. For example, some non-distro app may need to be installed. Or an older version of a supported library. Whatever. Standardising the container description doesn't mean they are all identical, each program will have it's different requirements and @GiancarloFusiello will need the help of the program authors when changes are needed. 

There is this traditional barrier between system admins and developers that the practice of devops was supposedly going to eliminate, however the division between system concerns and software concerns is still a natural division. When the tooling allows the developer to modify the system (container) their code runs in then this hugely empowering, even if there is a layer of approval and standardisation they must go through.